src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs

```diff
@@ -197,66 +217,41 @@ public BlittableOperation BlittableOperation
         /// <typeparam name="T"></typeparam>
         /// <param name="instance">The instance.</param>
         /// <returns></returns>
-        public Dictionary<string, string> GetMetadataFor<T>(T instance)
+        public IDictionary<string, string> GetMetadataFor<T>(T instance)
         {
-            var documentInfo = GetDocumentMetadata(instance);
+            var documentInfo = GetDocumentInfo(instance);
+
+            if (documentInfo.MetadataInstance != null)
+                return documentInfo.MetadataInstance;
+
             var metadataAsBlittable = documentInfo.Metadata;
-            Dictionary<string, string> metadata = new Dictionary<string, string>();
-            var indexes = metadataAsBlittable.GetPropertiesByInsertionOrder();
-            foreach (var index in indexes)
-            {
-                var prop = metadataAsBlittable.GetPropertyByIndex(index);
-                metadata[prop.Item1] = prop.Item2.ToString();
-            }
-            ;
-            _entitiesWithMetadataInstance.Add(documentInfo.Entity);
+            var metadata = new MetadataAsDictionary(metadataAsBlittable);
+           _entitiesWithMetadataInstance.Add(documentInfo.Entity);
```

Comment:
You are already scanning through all the entities in the session, thee is no need to add another property that you would scan separately.
Do it in the `PrepareForEntitiesPuts`